spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryTests.java

```diff
@@ -0,0 +1,268 @@
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerMappingContext;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerTable;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.repository.query.QueryMethod;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Chengyuan Zhao
+ */
+public class SpannerStatementQueryTests {
+
+	@Test
+	public void compoundNameConventionTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullAndTraderIdLikeAndPriceTrueAndPriceFalse"
+						+ "AndPriceGreaterThanAndPriceLessThanEqualOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[]{
+		"BUY",
+		"abcd",
+		"abc123",
+		8.88,
+		3.33,
+		"ignored",
+		"ignored",
+		"blahblah",
+		"ignored",
+		"ignored",
+		1.11,
+		2.22,
+		};
+
+		when(spannerOperations.find(any(), (Statement) any(), any()))
+				.thenAnswer(invocation -> {
+					Statement statement = invocation.getArgument(1);
+					assertEquals(
+							"SELECT DISTINCT * FROM trades WHERE ( action=@tag0 AND ticker=@tag1 ) OR "
+									+ "( trader_id=@tag2 AND price<@tag3 ) OR ( price>=@tag4 AND id<>NULL AND "
+									+ "trader_id=NULL AND trader_id LIKE %@tag7 AND price=TRUE AND price=FALSE AND "
+									+ "price>@tag10 AND price<=@tag11 )ORDER BY id DESC LIMIT 3; {tag0: BUY, tag1:"
+									+ " abcd, tag8: ignored, tag9: ignored, tag10: 1.11, tag6: ignored, tag11: 2.22, "
+									+ "tag7: blahblah, tag4: 3.33, tag5: ignored, tag2: abc123, tag3: 8.88}",
+							statement.toString());
+					return null;
+				});
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void unspecifiedParametersTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+		};
+
+		partTreeSpannerQuery.execute(params);
```

Comment:
readability/semantics: it is unclear what the unspecified parameter is, please restructure

Reply:
Could you elaborate? We can only give params as mysterious array of objects.

Reply:
sorry, I wasn't clear: what are we testing?

Reply:
There are too few params for the method name that was specified. 
"unspecifiedParameters" means there were some left ungiven.